src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1412,68 +1410,19 @@ TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTre
  * shard-extended deparsed SQL to be run during execution.
  */
 static Job *
-RouterInsertJob(Query *originalQuery, Query *query, DeferredErrorMessage **planningError)
+RouterInsertJob(Query *originalQuery, Query *query)
 {
-	Oid distributedTableId = ExtractFirstDistributedTableId(query);
-	List *taskList = NIL;
-	bool requiresMasterEvaluation = false;
-	bool deferredPruning = false;
-	Const *partitionKeyValue = NULL;
-
 	bool isMultiRowInsert = IsMultiRowInsert(query);
 	if (isMultiRowInsert)
 	{
 		/* add default expressions to RTE_VALUES in multi-row INSERTs */
 		NormalizeMultiRowInsertTargetList(originalQuery);
 	}
 
-	if (isMultiRowInsert || !CanShardPrune(distributedTableId, query))
-	{
-		/*
-		 * If there is a non-constant (e.g. parameter, function call) in the partition
-		 * column of the INSERT then we defer shard pruning until the executor where
-		 * these values are known.
-		 *
-		 * XXX: We also defer pruning for multi-row INSERTs because of some current
-		 * limitations with the way multi-row INSERTs are handled. Most notably, we
-		 * don't evaluate functions in task->rowValuesList. Therefore we need to
-		 * perform function evaluation before we can run RouterInsertTaskList.
-		 */
-		taskList = NIL;
-		deferredPruning = true;
-
-		/* must evaluate the non-constant in the partition column */
-		requiresMasterEvaluation = true;
-	}
-	else
-	{
-		bool parametersInQueryResolved = false;
-
-		taskList = RouterInsertTaskList(query, parametersInQueryResolved, planningError);
-		if (*planningError)
-		{
-			return NULL;
-		}
-
-		/* determine whether there are function calls to evaluate */
-		requiresMasterEvaluation = RequiresMasterEvaluation(originalQuery);
-	}
-
 	Job *job = CreateJob(originalQuery);
-	job->taskList = taskList;
-	job->requiresMasterEvaluation = requiresMasterEvaluation;
-	job->deferredPruning = deferredPruning;
-
-	if (!requiresMasterEvaluation)
-	{
-		/* no functions or parameters, build the query strings upfront */
-		RebuildQueryStrings(job);
-
-		/* remember the partition column value */
-		partitionKeyValue = ExtractInsertPartitionKeyValue(originalQuery);
-	}
-
-	job->partitionKeyValue = partitionKeyValue;
+	job->requiresMasterEvaluation = RequiresMasterEvaluation(originalQuery);
+	job->deferredPruning = true;
+	job->partitionKeyValue = ExtractInsertPartitionKeyValue(originalQuery);
```

Comment:
We already do this in `RegenerateTaskListForInsert `, so probably doesn't need to do it here? citus-stat-statements requires it in `EndScan`, which is fine.

Or maybe `IsLoggable(DEBUG2)` to keep the test unchanged and make it easier to debug for users. But, still my slight preference is to remove it all

Reply:
We skip function evaluation if the partition key value is known at this stage and there are no function calls, so there's some performance benefit to keeping it here.

Reply:
On the other hand, we call this function twice for `INSERT` without params/functions.

In either case, I don't expect any significant difference, so I'm fine with this.

Reply:
True, I'll add an `partitionKeyValue == NULL` check in RegenerateTaskList.